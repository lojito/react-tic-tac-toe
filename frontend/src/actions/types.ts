import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

export type CategoryAction = {
  type: "CHANGE_CATEGORY";
  category: Categories;
};

export type DisabledAction = {
  type: "CHANGE_DISABLED";
  disabled: boolean;
};

export type FirstAction = {
  type: "CHANGE_FIRST";
  first: FirstPlayer;
};

export type UserImageAction = {
  type: "CHANGE_USER_IMAGE";
  userImage: ImageName;
};

export type ComputerImageAction = {
  type: "CHANGE_COMPUTER_IMAGE";
  computerImage: ImageName;
};

export type PlayingLevelAction = {
  type: "CHANGE_PLAYING_LEVEL";
  level: PlayingLevel;
};

export type GameOverAction = {
  type: "CHANGE_GAME_OVER";
  over: boolean;
};

export type ResultAction = {
  type: "CHANGE_GAME_RESULT";
  result: GameResult;
};

export type WinnersSquaresAction = {
  type: "CHANGE_WINNERS_SQUARES";
  winners: SquareLocation[];
};

export type ActionUnknow = {
  type: "UNKNOWN";
};

export type Action =
  | CategoryAction
  | DisabledAction
  | FirstAction
  | UserImageAction
  | ComputerImageAction
  | PlayingLevelAction
  | GameOverAction
  | ResultAction
  | WinnersSquaresAction
  | ActionUnknow;
