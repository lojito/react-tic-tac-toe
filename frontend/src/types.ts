export type CategoryId = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

export type CategoryHavanaLandmarks = {
  id: 0;
  name: "CATEGORY_HAVANA_LANDMARKS";
  folder: "habana";
};

export type CategoryMontrealLandmarks = {
  id: 1;
  name: "CATEGORY_MONTREAL_LANDMARKS";
  folder: "montreal";
};

export type CategoryVancouverLandmarks = {
  id: 2;
  name: "CATEGORY_VANCOUVER_LANDMARKS";
  folder: "vancouver";
};

export type CategorySpainLandmarks = {
  id: 3;
  name: "CATEGORY_SPAIN_LANDMARKS";
  folder: "spain";
};

export type CategoryGermanyLandmarks = {
  id: 4;
  name: "CATEGORY_GERMANY_LANDMARKS";
  folder: "germany";
};

export type CategorySoccerPlayers = {
  id: 5;
  name: "CATEGORY_SOCCER_PLAYERS";
  folder: "soccer";
};

export type CategoryFruitsAndVegetables = {
  id: 6;
  name: "CATEGORY_FRUITS_AND_VEGETABLES";
  folder: "fruits";
};

export type CategoryAnimals = {
  id: 7;
  name: "CATEGORY_ANIMALS";
  folder: "animals";
};

export type CategoryPuppies = {
  id: 8;
  name: "CATEGORY_PUPPIES";
  folder: "puppies";
};

export type CategorySeinfeld = {
  id: 9;
  name: "CATEGORY_SEINFELD";
  folder: "seinfeld";
};

export type Categories =
  | CategoryAnimals
  | CategoryFruitsAndVegetables
  | CategoryGermanyLandmarks
  | CategoryHavanaLandmarks
  | CategoryMontrealLandmarks
  | CategoryPuppies
  | CategorySeinfeld
  | CategorySoccerPlayers
  | CategorySpainLandmarks
  | CategoryVancouverLandmarks;

export type ImageFolder =
  | "habana"
  | "montreal"
  | "vancouver"
  | "spain"
  | "germany"
  | "soccer"
  | "fruits"
  | "animals"
  | "puppies"
  | "seinfeld";

export type ImageName =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19;

export type FirstPlayer = Player.User | Player.Computer;

export enum PlayingLevel {
  Easy,
  Smart,
}

export enum Player {
  User,
  Computer,
  Nobody,
}

export type GameBoard = [
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player
];

export type SquareBoard = {
  disabled: boolean;
  id: string;
  imagePath: string;
  className: "win" | "";
};

export enum SquareLocation {
  TopLeft,
  TopCenter,
  TopRight,
  MiddleLeft,
  MiddleCenter,
  MiddleRight,
  BottomLeft,
  BottomCenter,
  BottomRight,
}

export enum GameResult {
  Draw,
  Lost,
  Started,
  Won,
}

export type DBGame = {
  _id?: string;
  board: GameBoard;
  categoryId: CategoryId;
  first: FirstPlayer;
  userImage: ImageName;
  computerImage: ImageName;
  level: PlayingLevel;
  result: GameResult;
  winners: SquareLocation[];
  user?: string;
  createdAt?: string;
  updatedAt?: string;
  __v?: number;
};
