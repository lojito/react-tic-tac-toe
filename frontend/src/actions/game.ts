import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

import {
  CategoryAction,
  ComputerImageAction,
  DisabledAction,
  FirstAction,
  GameOverAction,
  PlayingLevelAction,
  ResultAction,
  UserImageAction,
  WinnersSquaresAction,
} from "./types";

export const CHANGE_CATEGORY = "CHANGE_CATEGORY";
export const CHANGE_DISABLED = "CHANGE_DISABLED";
export const CHANGE_FIRST = "CHANGE_FIRST";
export const CHANGE_IMAGES = "CHANGE_IMAGES";
export const CHANGE_USER_IMAGE = "CHANGE_USER_IMAGE";
export const CHANGE_COMPUTER_IMAGE = "CHANGE_COMPUTER_IMAGE";
export const CHANGE_PLAYING_LEVEL = "CHANGE_PLAYING_LEVEL";
export const CHANGE_GAME_OVER = "CHANGE_GAME_OVER";
export const CHANGE_GAME_RESULT = "CHANGE_GAME_RESULT";
export const CHANGE_WINNERS_SQUARES = "CHANGE_WINNERS_SQUARES";

export const changeCategory = (category: Categories): CategoryAction => {
  return {
    type: CHANGE_CATEGORY,
    category,
  };
};

export const changeDisabled = (disabled: boolean): DisabledAction => {
  return {
    type: CHANGE_DISABLED,
    disabled,
  };
};

export const changeFirst = (first: FirstPlayer): FirstAction => {
  return {
    type: CHANGE_FIRST,
    first,
  };
};

export const changeUserImage = (userImage: ImageName): UserImageAction => {
  return {
    type: CHANGE_USER_IMAGE,
    userImage,
  };
};

export const changeComputerImage = (
  computerImage: ImageName
): ComputerImageAction => {
  return {
    type: CHANGE_COMPUTER_IMAGE,
    computerImage,
  };
};

export const changeLevel = (level: PlayingLevel): PlayingLevelAction => {
  return {
    type: CHANGE_PLAYING_LEVEL,
    level,
  };
};

export const changeOver = (over: boolean): GameOverAction => {
  return {
    type: CHANGE_GAME_OVER,
    over,
  };
};

export const changeResult = (result: GameResult): ResultAction => {
  return {
    type: CHANGE_GAME_RESULT,
    result,
  };
};

export const changeWinners = (
  winners: SquareLocation[]
): WinnersSquaresAction => {
  return {
    type: CHANGE_WINNERS_SQUARES,
    winners,
  };
};
