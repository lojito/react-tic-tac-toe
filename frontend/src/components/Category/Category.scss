.categories {
  text-align: center;
  font-weight: bold;
  margin-bottom: 5px;
}
