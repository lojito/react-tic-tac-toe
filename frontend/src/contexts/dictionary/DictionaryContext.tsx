import { createContext } from "react";
import { Language } from "../../contexts/language/LanguageContext";
import data from "../../data/json/dictionary.json";

export type DictionaryEntry = Partial<
  Record<
    | "ABOUT"
    | "ABOUT_GAME_DEVELOPED_WITH"
    | "ABOUT_OTHER_HOME_PROJECTS"
    | "ABOUT_SOURCE_CODE"
    | "AUTH_FORM_EMAIL_LABEL"
    | "AUTH_FORM_EMAIL_PLACEHOLDER"
    | "AUTH_FORM_NAME_LABEL"
    | "AUTH_FORM_NAME_PLACEHOLDER"
    | "AUTH_FORM_PASSWORD_LABEL"
    | "AUTH_FORM_PASSWORD_PLACEHOLDER"
    | "AUTH_ERROR_SERVER_NOT_RESPONDING"
    | "AUTH_ERROR_VALIDATION_FAILED"
    | "AUTH_ERROR_UNKNOWN"
    | "BUTTON_BACK"
    | "BUTTON_DELETE_GAME"
    | "BUTTON_LOGOUT"
    | "BUTTON_PLAY"
    | "BUTTON_PLAY_AGAIN"
    | "BUTTON_SEE_GAME"
    | "BUTTON_SIGNIN"
    | "BUTTON_SIGNUP"
    | "CATEGORY"
    | "CATEGORY_ANIMALS"
    | "CATEGORY_FRUITS_AND_VEGETABLES"
    | "CATEGORY_GERMANY_LANDMARKS"
    | "CATEGORY_HAVANA_LANDMARKS"
    | "CATEGORY_MONTREAL_LANDMARKS"
    | "CATEGORY_PUPPIES"
    | "CATEGORY_SEINFELD"
    | "CATEGORY_SOCCER_PLAYERS"
    | "CATEGORY_SPAIN_LANDMARKS"
    | "CATEGORY_VANCOUVER_LANDMARKS"
    | "CONTACT"
    | "CONTACT_AUTHOR"
    | "CONTACT_EMAIL"
    | "CONTACT_GITHUB"
    | "CONTACT_LINKEDIN"
    | "FLAG_ENGLISH_HINT"
    | "FLAG_FRENCH_HINT"
    | "FLAG_SPANISH_HINT"
    | "GAME"
    | "GAME_OVER"
    | "GAME_STARTED_BY"
    | "GAME_STARTED_BY_COMPUTER"
    | "GAME_STARTED_BY_YOU"
    | "GAME_DELETING"
    | "GAME_DELETE_ERROR"
    | "GAME_SAVE_ERROR"
    | "GAME_SAVING"
    | "HISTORICAL"
    | "HISTORICAL_NO_DATA_YET"
    | "HISTORICAL_FETCHING_GAMES"
    | "HISTORICAL_FETCHING_GAME"
    | "HISTORICAL_FETCHING_GAMES_ERROR"
    | "HISTORICAL_FETCHING_GAME_ERROR"
    | "IMAGE_COMPUTER_LABEL"
    | "IMAGE_COMPUTER_TOOLTIP"
    | "IMAGE_REFRESH_TOOLTIP"
    | "IMAGE_USER_LABEL"
    | "IMAGE_USER_TOOLTIP"
    | "LOGIN"
    | "MESSAGE_DRAW"
    | "MESSAGE_LOST"
    | "MESSAGE_WON"
    | "PLAYING_LEVEL_EASY"
    | "PLAYING_LEVEL_LABEL"
    | "PLAYING_LEVEL_SMART"
    | "PLAYING_LEVEL_TOOLTIP"
    | "REPOSITORY"
    | "SIGNUP"
    | "WELCOME_BACK"
    | "START_GAME_LABEL"
    | "START_GAME_ANSWER_NO"
    | "START_GAME_TOOLTIP"
    | "START_GAME_ANSWER_YES",
    string
  >
>;

export type DictionaryContextValue = {
  dictionary: DictionaryEntry;
};

export const DictionaryContext = createContext<DictionaryContextValue>({
  dictionary: data[Language.English],
});
