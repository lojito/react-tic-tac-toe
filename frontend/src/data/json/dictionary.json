[
  {
    "ABOUT": "About",
    "ABOUT_GAME_DEVELOPED_WITH": "Game developed with:",
    "ABOUT_OTHER_HOME_PROJECTS": "Other projects:",
    "ABOUT_SOURCE_CODE": "Source Code:",
    "AUTH_FORM_EMAIL_LABEL": "E-mail *",
    "AUTH_FORM_EMAIL_PLACEHOLDER": "E-mail",
    "AUTH_FORM_NAME_LABEL": "Name *",
    "AUTH_FORM_NAME_PLACEHOLDER": "Name",
    "AUTH_FORM_PASSWORD_LABEL": "Password *",
    "AUTH_FORM_PASSWORD_PLACEHOLDER": "Password",
    "AUTH_ERROR_SERVER_NOT_RESPONDING": "Server is not responding.",
    "AUTH_ERROR_VALIDATION_FAILED": "Validation failed. Invalid e-mail or password.",
    "AUTH_ERROR_UNKNOWN": "There was an error while trying to login or signup in the user.",
    "BUTTON_BACK": "Back",
    "BUTTON_DELETE_GAME": "Delete Game",
    "BUTTON_LOGOUT": "Logout",
    "BUTTON_PLAY": "Play Tic-Tac-Toe",
    "BUTTON_PLAY_AGAIN": "Play again",
    "BUTTON_SEE_GAME": "See Game",
    "BUTTON_SIGNIN": "Log in",
    "BUTTON_SIGNUP": "Sign up",
    "CATEGORY": "Category",
    "CATEGORY_ANIMALS": "Animals",
    "CATEGORY_FRUITS_AND_VEGETABLES": "Fruits & Vegetables",
    "CATEGORY_GERMANY_LANDMARKS": "Germany landmarks",
    "CATEGORY_HAVANA_LANDMARKS": "Havana landmarks",
    "CATEGORY_MONTREAL_LANDMARKS": "Montreal landmarks",
    "CATEGORY_PUPPIES": "Puppies",
    "CATEGORY_SEINFELD": "Seinfeld",
    "CATEGORY_SOCCER_PLAYERS": "Soccer players",
    "CATEGORY_SPAIN_LANDMARKS": "Spain landmarks",
    "CATEGORY_VANCOUVER_LANDMARKS": "Vancouver landmarks",
    "CONTACT": "Contact",
    "CONTACT_AUTHOR": "Author:",
    "CONTACT_EMAIL": "E-mail:",
    "CONTACT_GITHUB": "Github:",
    "CONTACT_LINKEDIN": "LinkedIn:",
    "FLAG_ENGLISH_HINT": "English",
    "FLAG_FRENCH_HINT": "French",
    "FLAG_SPANISH_HINT": "Spanish",
    "GAME": "Game",
    "GAME_OVER": "Game over",
    "GAME_STARTED_BY": "Started by",
    "GAME_STARTED_BY_COMPUTER": "Computer",
    "GAME_STARTED_BY_YOU": "You",
    "GAME_DELETING": "Removing the game data from the database...",
    "GAME_DELETE_ERROR": "An error ocurred while deleting the game from the database.",
    "GAME_SAVING": "Saving the game data...",
    "GAME_SAVE_ERROR": "An error ocurred while saving the game results into the database.",
    "HISTORICAL": "Historical",
    "HISTORICAL_NO_DATA_YET": "There is no historical data yet. Play some games and comeback.",
    "HISTORICAL_FETCHING_GAMES": "Fetching games...",
    "HISTORICAL_FETCHING_GAME": "Fetching game...",
    "HISTORICAL_FETCHING_GAMES_ERROR": "An error ocurred while fetching the games from the database.",
    "HISTORICAL_FETCHING_GAME_ERROR": "An error ocurred while fetching the game from the database.",
    "IMAGE_COMPUTER_LABEL": "PC",
    "IMAGE_COMPUTER_TOOLTIP": "The computer will play with this image.",
    "IMAGE_REFRESH_TOOLTIP": "Click to play with a different image.",
    "IMAGE_USER_LABEL": "You",
    "IMAGE_USER_TOOLTIP": "You will play with this image.",
    "LOGIN": "Log in",
    "MESSAGE_DRAW": "It's a draw!",
    "MESSAGE_LOST": "You lost!",
    "MESSAGE_WON": "You won!",
    "PLAYING_LEVEL_EASY": "Easy",
    "PLAYING_LEVEL_LABEL": "Level",
    "PLAYING_LEVEL_SMART": "Smart",
    "PLAYING_LEVEL_TOOLTIP": "How much thinking the computer does before playing.",
    "REPOSITORY": "Repository",
    "SIGNUP": "Sign up",
    "WELCOME_BACK": "Welcome back",
    "START_GAME_LABEL": "Start game",
    "START_GAME_ANSWER_NO": "No",
    "START_GAME_TOOLTIP": "Who starts playing the game: You or the computer?",
    "START_GAME_ANSWER_YES": "Yes"
  },
  {
    "ABOUT": "Acerca de",
    "ABOUT_GAME_DEVELOPED_WITH": "Juego desarrollado con:",
    "ABOUT_OTHER_HOME_PROJECTS": "Otros projectos:",
    "ABOUT_SOURCE_CODE": "Código fuente:",
    "AUTH_FORM_EMAIL_LABEL": "Correo electrónico *",
    "AUTH_FORM_EMAIL_PLACEHOLDER": "Correo electrónico",
    "AUTH_FORM_NAME_LABEL": "Nombre *",
    "AUTH_FORM_NAME_PLACEHOLDER": "Nombre",
    "AUTH_FORM_PASSWORD_LABEL": "Contraseña *",
    "AUTH_FORM_PASSWORD_PLACEHOLDER": "Contraseña",
    "AUTH_ERROR_SERVER_NOT_RESPONDING": "El servidor no está respondiendo.",
    "AUTH_ERROR_VALIDATION_FAILED": "Validación fallida. Correo electrónico o contraseña no válida.",
    "AUTH_ERROR_UNKNOWN": "Se produjo un error al intentar iniciar sesión o registrar al usuario.",
    "BUTTON_BACK": "Regresar",
    "BUTTON_DELETE_GAME": "Eliminar la partida",
    "BUTTON_LOGOUT": "Cerrar sesión",
    "BUTTON_PLAY": "Juega al Tic-Tac-Toe",
    "BUTTON_PLAY_AGAIN": "Jugar de nuevo",
    "BUTTON_SEE_GAME": "Ver la partida",
    "BUTTON_SIGNIN": "Iniciar sesión",
    "BUTTON_SIGNUP": "Inscribirse",
    "CATEGORY": "Categoría",
    "CATEGORY_ANIMALS": "Animales",
    "CATEGORY_FRUITS_AND_VEGETABLES": "Frutas y vegetales",
    "CATEGORY_GERMANY_LANDMARKS": "Monumentos de Alemania",
    "CATEGORY_HAVANA_LANDMARKS": "Monumentos de La Habana",
    "CATEGORY_MONTREAL_LANDMARKS": "Monumentos de Montreal",
    "CATEGORY_PUPPIES": "Cachorros",
    "CATEGORY_SEINFELD": "Seinfeld",
    "CATEGORY_SOCCER_PLAYERS": "Jugadores de fútbol",
    "CATEGORY_SPAIN_LANDMARKS": "Monumentos de España",
    "CATEGORY_VANCOUVER_LANDMARKS": "Monumentos de Vancouver",
    "CONTACT": "Contacto",
    "CONTACT_AUTHOR": "Autor:",
    "CONTACT_EMAIL": "Correo electrónico:",
    "CONTACT_GITHUB": "Github:",
    "CONTACT_LINKEDIN": "LinkedIn:",
    "FLAG_ENGLISH_HINT": "Inglés",
    "FLAG_FRENCH_HINT": "Francés",
    "FLAG_SPANISH_HINT": "Español",
    "GAME": "Juego",
    "GAME_OVER": "Partida acabada",
    "GAME_STARTED_BY": "Comenzada por",
    "GAME_STARTED_BY_COMPUTER": "Ordenador",
    "GAME_STARTED_BY_YOU": "Ti",
    "GAME_DELETING": "Eliminando el juego de la base de datos...",
    "GAME_DELETE_ERROR": "Ocurrió un error al eliminar el juego de la base de datos.",
    "GAME_SAVING": "Guardando los datos del juego...",
    "GAME_SAVE_ERROR": "Ocurrió un error al salvar los resultados del juego de la base de datos.",
    "HISTORICAL": "Histórico",
    "HISTORICAL_NO_DATA_YET": "No hay historial aún. Juega algunas partidas y regresa más tarde.",
    "HISTORICAL_FETCHING_GAMES": "Obteniendo juegos...",
    "HISTORICAL_FETCHING_GAME": "Obteniendo juego...",
    "HISTORICAL_FETCHING_GAMES_ERROR": "Ocurrió un error al recuperar los juegos de la base de datos.",
    "HISTORICAL_FETCHING_GAME_ERROR": "Ocurrió un error al recuperar el juego de la base de datos.",
    "IMAGE_COMPUTER_LABEL": "PC",
    "IMAGE_COMPUTER_TOOLTIP": "El ordenador jugará con esta imagen.",
    "IMAGE_REFRESH_TOOLTIP": "Haz click para jugar con una imagen diferente.",
    "IMAGE_USER_LABEL": "Tú",
    "IMAGE_USER_TOOLTIP": "Tú jugaras con esta imagen.",
    "LOGIN": "Iniciar sesión",
    "MESSAGE_DRAW": "Empate!",
    "MESSAGE_LOST": "Perdiste!",
    "MESSAGE_WON": "Ganaste!",
    "PLAYING_LEVEL_EASY": "Fácil",
    "PLAYING_LEVEL_LABEL": "Nivel de juego",
    "PLAYING_LEVEL_SMART": "Difícil",
    "PLAYING_LEVEL_TOOLTIP": "Cuánto tiempo el ordenador piensa antes de jugar.",
    "REPOSITORY": "Banco",
    "SIGNUP": "Regístrate",
    "WELCOME_BACK": "Bienvenido de nuevo",
    "START_GAME_LABEL": "Tú empiezas",
    "START_GAME_ANSWER_NO": "No",
    "START_GAME_TOOLTIP": "Quién empieza la partida: Tú o el ordenador?",
    "START_GAME_ANSWER_YES": "Sí"
  },
  {
    "ABOUT": "À propos de",
    "ABOUT_GAME_DEVELOPED_WITH": "Jeu développé avec:",
    "ABOUT_OTHER_HOME_PROJECTS": "D'autres projects:",
    "ABOUT_SOURCE_CODE": "Code source:",
    "AUTH_FORM_EMAIL_LABEL": "Courrier électronique *",
    "AUTH_FORM_EMAIL_PLACEHOLDER": "Courrier électronique",
    "AUTH_FORM_NAME_LABEL": "Nom *",
    "AUTH_FORM_NAME_PLACEHOLDER": "Nom",
    "AUTH_FORM_PASSWORD_LABEL": "Mot de passe *",
    "AUTH_FORM_PASSWORD_PLACEHOLDER": "Mot de passe",
    "AUTH_ERROR_SERVER_NOT_RESPONDING": "Le serveur ne répond pas.",
    "AUTH_ERROR_VALIDATION_FAILED": "Validation échouée. Email ou mot de passe invalide.",
    "AUTH_ERROR_UNKNOWN": "Une erreur s'est produite lors de la tentative de connexion ou d'inscription de l'utilisateur.",
    "BUTTON_BACK": "Retour",
    "BUTTON_DELETE_GAME": "Retirer la partie",
    "BUTTON_LOGOUT": "Se déconnecter",
    "BUTTON_PLAY": "Joue au Tic-Tac-Toe",
    "BUTTON_PLAY_AGAIN": "Jouer à nouveau",
    "BUTTON_SEE_GAME": "Voir la partie",
    "BUTTON_SIGNIN": "Se connecter",
    "BUTTON_SIGNUP": "S'inscrire",
    "CATEGORY": "Catégorie",
    "CATEGORY_ANIMALS": "Des animaux",
    "CATEGORY_FRUITS_AND_VEGETABLES": "Fruits et légumes",
    "CATEGORY_GERMANY_LANDMARKS": "Monuments L'Allemagne",
    "CATEGORY_HAVANA_LANDMARKS": "Monuments de La Havane",
    "CATEGORY_MONTREAL_LANDMARKS": "Monuments de Montréal",
    "CATEGORY_PUPPIES": "Des chiots",
    "CATEGORY_SEINFELD": "Seinfeld",
    "CATEGORY_SOCCER_PLAYERS": "Joueurs du soccer",
    "CATEGORY_SPAIN_LANDMARKS": "Monuments L'Espagne",
    "CATEGORY_VANCOUVER_LANDMARKS": "Monuments de Vancouver",
    "CONTACT": "Contact",
    "CONTACT_AUTHOR": "Auteur:",
    "CONTACT_EMAIL": "Courrier électronique:",
    "CONTACT_GITHUB": "Github:",
    "CONTACT_LINKEDIN": "LinkedIn:",
    "FLAG_ENGLISH_HINT": "Anglais",
    "FLAG_FRENCH_HINT": "Français",
    "FLAG_SPANISH_HINT": "Espagnol",
    "GAME": "Jeu",
    "GAME_OVER": "La partie est finie",
    "GAME_STARTED_BY": "Commencé par",
    "GAME_STARTED_BY_COMPUTER": "Ordinateur",
    "GAME_STARTED_BY_YOU": "Toi",
    "GAME_DELETING": "En train de supprimer le jeu de la bsae de données...",
    "GAME_DELETE_ERROR": "Une erreur s'est produite lors de la suppression du jeu de la base de données.",
    "GAME_SAVING": "En train de sauvegarder les données du jeu...",
    "GAME_SAVE_ERROR": "Une erreur s'est produite lors de l'enregistrement des résultats du jeu dans la base de données.",
    "HISTORICAL": "Historique",
    "HISTORICAL_NO_DATA_YET": "Il n'y a pas encore d'historique. Reviens un peu plus tard après avoir joué quelques parties.",
    "HISTORICAL_FETCHING_GAMES": "Récupération des jeux...",
    "HISTORICAL_FETCHING_GAME": "Récupération du jeu...",
    "HISTORICAL_FETCHING_GAMES_ERROR": "Une erreur s'est produite lors de la récupération des jeux de la base de données.",
    "HISTORICAL_FETCHING_GAME_ERROR": "Une erreur s'est produite lors de la récupération du jeu de la base de données.",
    "IMAGE_COMPUTER_LABEL": "Ordi",
    "IMAGE_COMPUTER_TOOLTIP": "L'ordinateur va jouer avec cette image.",
    "IMAGE_REFRESH_TOOLTIP": "Fais clic à fin de jouer avec une autre image.",
    "IMAGE_USER_LABEL": "Toi",
    "IMAGE_USER_TOOLTIP": "Tu vas jouer avec cette image.",
    "LOGIN": "Se connecter",
    "MESSAGE_DRAW": "Match nul!",
    "MESSAGE_LOST": "Tu as perdu!",
    "MESSAGE_WON": "Tu as gagné!",
    "PLAYING_LEVEL_EASY": "Facile",
    "PLAYING_LEVEL_LABEL": "Niveau de jeu",
    "PLAYING_LEVEL_SMART": "Difficile",
    "PLAYING_LEVEL_TOOLTIP": "Quelle est la réflexion de l'ordinateur avant de jouer.",
    "REPOSITORY": "Banque",
    "SIGNUP": "S'inscrire",
    "WELCOME_BACK": "Ravis de vous retrouver",
    "START_GAME_LABEL": "Premier",
    "START_GAME_ANSWER_NO": "Non",
    "START_GAME_TOOLTIP": "Qui commence la partie: L'ordinateur ou toi?",
    "START_GAME_ANSWER_YES": "Oui"
  }
]
