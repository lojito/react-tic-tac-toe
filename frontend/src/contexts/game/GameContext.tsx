import React, { createContext } from "react";
import { Action } from "../../actions/types";
import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../../types";

export type State = {
  category: Categories;
  userImage: ImageName;
  computerImage: ImageName;
  first: FirstPlayer;
  level: PlayingLevel;
  disabled: boolean;
  over: boolean;
  result: GameResult;
  winners: SquareLocation[];
};

export const GameContext = createContext<{
  game: State;
  dispatch: React.Dispatch<Action>;
}>({ game: {} as State, dispatch: {} as React.Dispatch<Action> });
