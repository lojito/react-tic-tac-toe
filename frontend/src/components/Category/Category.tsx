import React, { FC, memo, useContext } from "react";
import { changeCategory } from "../../actions/game";
import { Action } from "../../actions/types";
import {
  DictionaryContext,
  DictionaryEntry,
} from "../../contexts/dictionary/DictionaryContext";
import { GameContext } from "../../contexts/game/GameContext";
import categories from "../../data/json/categories.json";
import { Categories, CategoryId } from "../../types";
import "./Category.scss";

interface Props {
  disabled: boolean;
  categoryId: CategoryId;
  dispatch: React.Dispatch<Action>;
  categories: Categories[];
  dictionary: DictionaryEntry;
}

const MemoizedCategory: FC<Props> = memo(
  ({ disabled, categoryId, dispatch, categories, dictionary }) => {
    return (
      <div className="categories" data-testid="categories">
        <label htmlFor="categories">{dictionary.CATEGORY}: </label>
        <select
          id="categories"
          defaultValue={categoryId}
          onChange={(e) => {
            dispatch(changeCategory(categories[+e.target.value]));
          }}
          disabled={disabled}
        >
          {categories.map(({ id, name }: Categories) => {
            return (
              <option key={name} value={id}>
                {dictionary[name]}
              </option>
            );
          })}
        </select>
      </div>
    );
  }
);

const Category: FC = () => {
  const { dictionary } = useContext(DictionaryContext);
  const { game, dispatch } = useContext(GameContext);

  return (
    <MemoizedCategory
      disabled={game.disabled}
      categoryId={game.category.id}
      dispatch={dispatch}
      categories={categories as Categories[]}
      dictionary={dictionary}
    />
  );
};

export default Category;
