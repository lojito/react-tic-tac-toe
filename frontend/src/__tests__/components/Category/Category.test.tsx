import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Category from "../../../components/Category/Category";
import { DictionaryContext } from "../../../contexts/dictionary/DictionaryContext";
import { GameContext, State } from "../../../contexts/game/GameContext";
import { Categories } from "../../../types";

jest.mock("../../../actions/game", () => {
  return {
    changeCategory: (category: Categories) => {
      return {
        type: "CHANGE_CATEGORY",
        category,
      };
    },
  };
});

describe("Category", () => {
  const mockDictionary = {
    CATEGORY_ANIMALS: "Animals",
    CATEGORY_FRUITS_AND_VEGETABLES: "Fruits & Vegetables",
    CATEGORY_GERMANY_LANDMARKS: "Germany landmarks",
    CATEGORY_HAVANA_LANDMARKS: "Havana landmarks",
    CATEGORY_MONTREAL_LANDMARKS: "Montreal landmarks",
    CATEGORY_PUPPIES: "Puppies",
    CATEGORY_SEINFELD: "Seinfeld",
    CATEGORY_SOCCER_PLAYERS: "Soccer players",
    CATEGORY_SPAIN_LANDMARKS: "Spain landmarks",
    CATEGORY_VANCOUVER_LANDMARKS: "Vancouver landmarks",
    CATEGORY: "Category",
  };

  const renderCategory = () => {
    const game = {
      category: {
        id: 0,
        folder: "habana",
        name: "CATEGORY_HAVANA_LANDMARKS",
      },
    } as State;

    const dispatch = jest.fn();

    const utils = render(
      <DictionaryContext.Provider value={{ dictionary: mockDictionary }}>
        <GameContext.Provider value={{ game, dispatch }}>
          <Category />
        </GameContext.Provider>
      </DictionaryContext.Provider>
    );

    const divWrapper = screen.getByTestId("categories");

    const labelCategory = screen.getByLabelText(`${mockDictionary.CATEGORY}:`);

    const selectCategory = screen.getByRole("combobox");

    const optionVancouver = screen.getByRole("option", {
      name: mockDictionary.CATEGORY_VANCOUVER_LANDMARKS,
    }) as HTMLOptionElement;

    return {
      ...utils,
      mockDictionary,
      dispatch,
      divWrapper,
      labelCategory,
      selectCategory,
      optionVancouver,
    };
  };

  it("renders the Category component correctly", () => {
    const { asFragment } = renderCategory();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should apply the categories class", () => {
    const { divWrapper } = renderCategory();

    expect(divWrapper).toHaveClass("categories");
  });

  it("should display the category label", () => {
    const { labelCategory } = renderCategory();

    expect(labelCategory).toBeInTheDocument();
  });

  it("should change the current category", async () => {
    const user = userEvent.setup();
    const categoryVancouver = {
      folder: "vancouver",
      id: 2,
      name: "CATEGORY_VANCOUVER_LANDMARKS",
    };
    const { dispatch, selectCategory, optionVancouver } = renderCategory();

    await user.selectOptions(selectCategory, optionVancouver);

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({
      category: categoryVancouver,
      type: "CHANGE_CATEGORY",
    });
  });
});
