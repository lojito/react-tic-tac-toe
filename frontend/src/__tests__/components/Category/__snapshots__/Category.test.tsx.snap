// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category renders the Category component correctly 1`] = `
<DocumentFragment>
  <div
    class="categories"
    data-testid="categories"
  >
    <label
      for="categories"
    >
      Category: 
    </label>
    <select
      id="categories"
    >
      <option
        selected=""
        value="0"
      >
        Havana landmarks
      </option>
      <option
        value="1"
      >
        Montreal landmarks
      </option>
      <option
        value="2"
      >
        Vancouver landmarks
      </option>
      <option
        value="3"
      >
        Spain landmarks
      </option>
      <option
        value="4"
      >
        Germany landmarks
      </option>
      <option
        value="5"
      >
        Soccer players
      </option>
      <option
        value="6"
      >
        Fruits & Vegetables
      </option>
      <option
        value="7"
      >
        Animals
      </option>
      <option
        value="8"
      >
        Puppies
      </option>
      <option
        value="9"
      >
        Seinfeld
      </option>
    </select>
  </div>
</DocumentFragment>
`;
